File: dijkstra.pas
(* the shortest way with the path printed *)
uses
   strings;
const
   min_point = 1;
   max_point = 6;  (* at range 0..255 *)
   max_path_length = 1000;
   ARCS = 11; (* number of arcs *)
type
   range = min_point..max_point;
   bytes = set of range;
const
   ini_arcs: array[1..ARCS]of array[1..3]of integer =
      ((1,2,4),   (* start, end, length *)
       (1,3,11),
       (2,3,5),
       (2,4,2),
       (2,5,8),
       (2,6,7),
       (3,5,5),
       (3,6,3),
       (4,5,1),
       (4,6,10),
       (5,6,3));
   vertices: bytes = [];
   start_point = 1;
   end_point = 6;
var
   D: array[range] of integer;
   P: array[range] of PChar;
   ts: string;
   lengths: array[range, range] of integer;
   S: bytes;
   i, j, l: integer;
procedure fill_sets;
   var i, v1, v2: byte;
   begin
       for v1 := min_point to max_point do
          for v2 := min_point to max_point do
             lengths[v1][v2] := -1;
       for i := 0 to high(ini_arcs) do begin
          v1 := ini_arcs[i][1];
          v2 := ini_arcs[i][2];
          vertices := vertices + [v1] + [v2];
          lengths[v1][v2] := ini_arcs[i][3];
          lengths[v2][v1] := ini_arcs[i][3];
       end
   end;
function minD: integer;
   var
      min: integer;
      index_min: integer;
   begin
      min := high(integer);
      index_min := -1;
      for i := min_point to max_point do
         if i in S then
            if (D[i] >= 0) and (min > D[i]) then begin
               min := D[i];
               index_min := i
            end;
      minD := index_min
   end;
begin
   for i := min_point to  max_point do
      P[i] := StrAlloc(max_path_length);
   fill_sets;
   D[start_point] := 0;                         (* Step 1 *)
   S := vertices - [start_point];
   for i := min_point to max_point do           (* Step 2 *)
      if i in S then begin
         D[i] := lengths[start_point][i];
         str(i, ts);
         strpcopy(P[i], ts);
      end;
   repeat                                       (* Step 3 *)
      j := minD;
      S := S - [j];
      if j = end_point then begin               (* Step 4 *)
         writeln('min = ', D[j], ' ', start_point, '-', P[end_point]);
         halt
      end;
      for i := min_point to max_point do        (* Step 5 *)
         if i in S then begin
            l := lengths[j][i];
            if l >= 0 then begin
               l := D[j] + l;
               if (D[i] < 0) or (D[i] > l) then begin
                  D[i] := l;
                  str(i, ts);
                  strcopy(P[i], P[j]);
                  strcat(P[i], '-');
                  ts := ts + #0;
                  strcat(P[i], @ts[1])
               end;
            end
         end
   until false
end.

File: dijkstra.cpp
#define RANDOM_FILL 0
#define PRINT_BOX 0

#include <iostream>
#include <iomanip>
#include <set>
#include <string>
#include <climits>
#include <cstdlib>
#include <ctime>
using namespace std;

#if RANDOM_FILL
const unsigned start_point = 0, BOX_SIZE = 200, TOPS = BOX_SIZE*BOX_SIZE, end_point = TOPS - 1, max_arc_length = 9;
#else
const unsigned ARCS = 11, TOPS = 6, start_point = 0, end_point = 5;
unsigned ini_arcs[ARCS][3] = /* start, end, length */
       {{1,2,4},
       {1,3,11},
       {2,3,5},
       {2,4,2},
       {2,5,8},
       {2,6,7},
       {3,5,5},
       {3,6,3},
       {4,5,1},
       {4,6,10},
       {5,6,3}};
#endif

set<unsigned> S;
unsigned (*lengths)[TOPS], D[TOPS], arc_count[TOPS];
string path[TOPS]; //path printed

#if RANDOM_FILL
void fill_random_box() {
	unsigned v1, v2;
	srand(time(0)%777);
	for (v1 = 0; v1 < TOPS; ++v1) {
		S.insert(v1);
		for (v2 = 0; v2 < TOPS; ++v2)
			if (v1 < v2) {
				if (v1 + 1 == v2 && v2%BOX_SIZE != 0 || v1 + BOX_SIZE == v2)
					lengths[v2][v1] = lengths[v1][v2] = rand()%max_arc_length + 1;
				else
					lengths[v2][v1] = lengths[v1][v2] = UINT_MAX;
			}
	}
}
void print_box() {
	unsigned v1, v2, w1 = 1, w2 = 2;
/*	for (v1 = 0; v1 < TOPS; ++v1)
		for (v2 = 0; v2 < TOPS; ++v2)
			if (v1 != v2)
			cout << v1 << '-' << v2 << ' ' << lengths[v1][v2] << endl;*/
        v1 = TOPS;
	while ((v1 /= 10) > 0) ++w1;
        v1 = max_arc_length;
	while ((v1 /= 10) > 0) ++w2;
	for (v1 = 0; v1 < BOX_SIZE; ++v1) {
		for (v2 = 0; v2 < BOX_SIZE - 1; ++v2)
			cout << '[' << setw(w1) << v1*BOX_SIZE + v2 << ']' << setw(w2) 
				<< lengths[v1*BOX_SIZE + v2][v1*BOX_SIZE + v2 + 1];
		cout << '[' << setw(w1) << v1*BOX_SIZE + v2 << "]\n";
                if (v1 < BOX_SIZE - 1) {
			for (v2 = 0; v2 < BOX_SIZE - 1; ++v2)
				cout << setw(w2) << lengths[v1*BOX_SIZE + v2][v1*BOX_SIZE + v2 + BOX_SIZE]
					<< setw(w1 + 2) << " ";
			cout << setw(w2) << lengths[v1*BOX_SIZE + v2][v1*BOX_SIZE + v2 + BOX_SIZE] << endl;
		}
	}
}
#else
void fill_sets() {
	unsigned i, v1, v2;
	for (v1 = 0; v1 < TOPS; ++v1)
		for (v2 = 0; v2 < TOPS; ++v2)
			lengths[v1][v2] = UINT_MAX;
	for (i = 0; i < ARCS; ++i) {
		v1 = ini_arcs[i][0] - 1;
		v2 = ini_arcs[i][1] - 1;
		S.insert(v1);
		S.insert(v2);
		lengths[v1][v2] = ini_arcs[i][2];
		//lengths[v2][v1] = ini_arcs[i][2];
	}
}
#endif

unsigned minD() {
	unsigned min, index_min;
	min = UINT_MAX;
	for (auto i: S)
		if (min > D[i]) {
			min = D[i];
			index_min = i;
		}
	return index_min;
}
int main() {
	unsigned j, l;
	lengths = (unsigned(*)[TOPS]) new unsigned [TOPS*TOPS];
#if RANDOM_FILL
	fill_random_box();
#if PRINT_BOX
        print_box();
#endif
#else
	fill_sets();
#endif
	D[start_point] = 0;
	S.erase(start_point);			//Step 1
	for (auto i: S) {				//Step 2
		D[i] = lengths[start_point][i];
		path[i] = to_string(start_point) + "-" + to_string(i);
		arc_count[i]++;
	}
	for(;;) {					//Step 3
		j = minD();
		S.erase(j);
		if (j == end_point) {			//Step 4
			cout << "min = "  << D[j] << " (" << arc_count[end_point] << ")\npath = " << path[end_point] << endl;
			return 0;
		}
		for (auto i: S)				//Step 5
			if ((l = lengths[j][i]) != UINT_MAX) {
				l += D[j];
				if (D[i] > l) {
					D[i] = l;
					path[i] = path[j] + "-" + to_string(i);
					arc_count[i] = arc_count[j] + 1;
				}
			}
	}
}

